cEP-0021.md

```diff
@@ -0,0 +1,265 @@
+| Authors  | Vaibhav Kumar Rai <mailto:raivaibhav08@gmail.com> |
+| Status   | Proposed                                          |
+| Type     | Process                                           |
+
+## Abstract
+
+This cEP describes the detailed implementation of profiling interface as a part
+of the
+[GSoC'18 project](https://summerofcode.withgoogle.com/projects/#6109762077327360).
+
+## Introduction
+
+Profiling a Python program is doing a dynamic analysis that measures the
+execution time of the program and everything that compose it. That means
+measuring the time spent in each of its functions. This will give the data about
+where our program is spending time, and what area might be worth optimizing,
+thus we can provide an *Profiling interface* from within coala and make it
+available to bear writers so that they can create Better Performant bears.
+
+This proposal is to introduce the Profiling interface so that bear writer will
+have the ability to jump directly into Bear code and move through it using a
+debugger interface and Profile the Bear code to create Better Performant
+bears.
+
+## Proposed Change
+
+Bear writer can invoke Profiler through *CLI*:
+
+- By passing custom settings through *CLI*
+  `coala -b Bear Mybear --files file --profile-bears -"True,...."`,
+- By using coala default settings provided by coala
+  `coala -b Bear Mybear --files file --profile-bears`,the default are to strip
+  the directory (`strip_dirs()`) and sort the stats of profiler result by
+  cumulative time (`cumtime`) and display the profiler result on *CLI*
+
+Bear writer can also invoke the Profiler by using `coafile` i.e.
+`profile_bears = XXX`.
+
+- The profiling will start by profiling the `run()` method of the bears
+  The advantage of profiling just the `run()` method is that, bear writer
+  doesn't have to spend time on loading the files, collecting the settings,
+  etc. as this is all done by coala itself, they just have to Profile
+  the `run()` method to create a Better Performant bear.
+
+After invoking the profiler on the run method of a bear it will first save
+all the profile output on a local temp file which further used by
+`pstats module` to sort the profile results and formats them in a way desired
+by the Bear writer.
+
+> Currently default format is to strip the directory (`strip_dirs()`) and sort
+> the stats of profiler result by cummulative time (`cumtime`) and display the
+> profiler result on *CLI*.
+>
+> ```python
+> pstats.strip_dirs().sort_stats('cumtime')
+> ```
+
+Bear writer will have an option to dump the formatted result in a
+specified file, instead of displaying it on terminal. i.e. by using
+`filename/file-path` instead of `True`.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "<True or filename/file-path>"
+```
+
+Bear writer will also have a option to directly dump the profile output,
+which can be further use for estimation of profiler stats.
```

Comment:
- `an option`
- the `raw profile output` may be better.
- can be further `used`